ca/pkcs8/pkcs8.go

```diff
@@ -0,0 +1,273 @@
+	block, _ := pem.Decode(pemBytes)
+	if block == nil {
+		return nil, errors.New("invalid pem file")
+	}
+
+	var (
+		der []byte
+		err error
+	)
+	der = block.Bytes
+
+	if ok := IsEncryptedPEMBlock(block); ok {
+		der, err = DecryptPEMBlock(block, password)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return derhelpers.ParsePrivateKeyDER(der)
+}
+
+// IsEncryptedPEMBlock checks if a PKCS#8 PEM-block is encrypted or not
+func IsEncryptedPEMBlock(block *pem.Block) bool {
+	der := block.Bytes
+
+	var privKey encryptedPrivateKeyInfo
+	if _, err := asn1.Unmarshal(der, &privKey); err != nil {
+		return false
+	}
+
+	return true
+}
+
+// DecryptPEMBlock requires PKCS#8 PEM Block and password to decrypt and return unencrypted der []byte
+func DecryptPEMBlock(block *pem.Block, password []byte) ([]byte, error) {
+	der := block.Bytes
+
+	var privKey encryptedPrivateKeyInfo
+	if _, err := asn1.Unmarshal(der, &privKey); err != nil {
+		return nil, errors.New("pkcs8: only PKCS #5 v2.0 supported")
+	}
+
+	if !privKey.EncryptionAlgorithm.IDPBES2.Equal(oidPBES2) {
+		return nil, errors.New("pkcs8: only PBES2 supported")
+	}
+
+	if !privKey.EncryptionAlgorithm.PBES2Params.KeyDerivationFunc.IDPBKDF2.Equal(oidPKCS5PBKDF2) {
+		return nil, errors.New("pkcs8: only PBKDF2 supported")
+	}
+
+	encParam := privKey.EncryptionAlgorithm.PBES2Params.EncryptionScheme
+	kdfParam := privKey.EncryptionAlgorithm.PBES2Params.KeyDerivationFunc.PBKDF2Params
+
+	switch {
+	case encParam.EncryAlgo.Equal(oidAES256CBC):
+		iv := encParam.IV
+		salt := kdfParam.Salt
+		iter := kdfParam.IterationCount
+
+		encryptedKey := privKey.EncryptedData
+		symkey := pbkdf2.Key(password, salt, iter, 32, sha1.New)
+		block, err := aes.NewCipher(symkey)
+		if err != nil {
+			return nil, err
+		}
+		mode := cipher.NewCBCDecrypter(block, iv)
+		mode.CryptBlocks(encryptedKey, encryptedKey)
+
+		// Remove padding
+		keyLen := len(encryptedKey)
+		padLen := int(encryptedKey[keyLen-1])
+		if padLen > keyLen || padLen > aes.BlockSize {
+			return nil, errors.New("pkcs8: invalid padding size")
+		}
+		encryptedKey = encryptedKey[:keyLen-padLen]
+
+		if _, err := derhelpers.ParsePrivateKeyDER(encryptedKey); err != nil {
+			return nil, errors.New("pkcs8: incorrect password")
+		}
+
+		return encryptedKey, nil
+	default:
+		return nil, errors.New("pkcs8: only AES-256-CBC supported")
+	}
+}
+
+func encryptPrivateKey(pkey, password []byte) ([]byte, error) {
+	// Calculate key from password based on PKCS5 algorithm
+	// Use 8 byte salt, 16 byte IV, and 2048 iteration
+	iter := 2048
+	salt := make([]byte, 8)
+	iv := make([]byte, 16)
+
+	if _, err := rand.Reader.Read(salt); err != nil {
+		return nil, err
+	}
+
+	if _, err := rand.Reader.Read(iv); err != nil {
+		return nil, err
+	}
+
+	key := pbkdf2.Key(password, salt, iter, 32, sha1.New)
+
+	// Use AES256-CBC mode, pad plaintext with PKCS5 padding scheme
+	padding := aes.BlockSize - len(pkey)%aes.BlockSize
+	if padding > 0 {
```

Comment:
This code seems correct, but operating on the final key buffer directly when padding is asking for trouble.

- Let's define `n := len(pkey)` outside of the `if`
- Let's create the padding array first, fill it with `0..len(padding)` and only then we append the padding array to `pkey`